node/Dockerfile

```diff
@@ -0,0 +1,23 @@
+# Build Stage
+FROM rust:1.72.1 as build
+COPY . /node/
```

Comment:
this is a bad idea, because node/target contains the local build artifacts which can be enormous.

Reply:
it would be nice to have it rebuilt only if the code changes

Reply:
I'm not entirely sure if I grasp this correctly. I completely agree that copying the entire node crate into Docker could lead to problems. Would moving everything outside the `node` directory help alleviate this? Currently, the entire crate is built within a Docker image, and the final image only includes the generated binaries. I attempted to compile everything outside Docker and only copy the `target` directory, but that requires cross-compilation to x86_64 architecture if run on a Mac or another ARM machine. It's technically feasible, but I'm uncertain if we intend to pursue this, considering potential additional dependencies. Let me know if my understanding is accurate or if there's something I might be overlooking.

Reply:
I would recommend copying everything EXCEPT the target directory to avoid pulling in local artifacts. Alternatively you can try to pull in only files tracked by git (git ls-tree) but I don't know if that's an easy thing to do.

Reply:
moving your docker-related configs outside of the node directory will prevent rebuilding the image when you change the config. Ideally it would be great to have the "cargo build" command split into building dependencies and the local code, in a way that the image layer with dependencies' artifacts is not rebuilt every time you update the local code (I don't know how to do that though without changing the build system, so this is just a thought for the future).

Reply:
Okay got it. I added the `node/target` dir to the `.dockerignore` file so it should not be cloned inside the containers. Also I moved every docker related file and the makefile to the root of the project and adapted every script and command to work with the new paths. Let me know if I missed something

Reply:
Regarding incorporating dependencies and local code in different Docker layers to cache the initial ones, it seems feasible. However, as you mentioned, this would entail altering the structure of the `Cargo.toml`. My initial thought on achieving this is to include only external dependencies in the `Cargo.toml` of the `node` crate. Then, within the `Cargo.toml` of each distinct crate, build every member of the workspace. This way, we could initially build all the external dependencies to cache them and subsequently build the actual node code. I plan to explore this in a separate pull request to avoid making unnecessary changes for this one.